### PR TITLE
IGNITE-13358

### PR DESCRIPTION
**Pull request by ascherbakoff**
Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/preloader/GridDhtPreloader.java [hunk 1]
```diff
@@ -223,27 +223,8 @@ public boolean disableRebalancingCancellationOptimization() {
...
-                if (part.state() == EVICTED) {
-                    part.awaitDestroy();
-
-                    part = top.localPartition(p, topVer, true);
-
-                    assert part != null : "Partition was not created [grp=" + grp.name() + ", topVer=" + topVer + ", p=" + p + ']';
-
-                    part.resetUpdateCounter();
-                }
+                // State should be switched to MOVING (or partition recreated) during PME.
```
Comment (agoncharuk) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/preloader/GridDhtPreloader.java:226]: Let's have a single check for these states

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/topology/GridDhtPartitionTopologyImpl.java [hunk 2]
```diff
@@ -2848,25 +2795,55 @@ private void removeNode(UUID nodeId) {
...
+
+                part.clearDeferredDeletes();
+
+                List<GridDhtLocalPartition> parts = localPartitions();
+
+                int renting = 0;
+
+                for (GridDhtLocalPartition part0 : parts) {
+                    if (part0.state() == RENTING)
+                        renting++;
```
Comment (agoncharuk) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/topology/GridDhtPartitionTopologyImpl.java:2839]: Minor, but we can break the loop as soon as we found first RENTING partition, no need to keep iteration.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCacheConcurrentMapImpl.java [hunk 3]
```diff
@@ -95,7 +95,7 @@ protected final GridCacheMapEntry putEntryIfObsoleteOrAbsent(
         GridCacheMapEntry doomed = null;
 
         boolean done = false;
-        boolean reserved = false;
+        boolean reserved = clearing;
```
Comment (agoncharuk) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/GridCacheConcurrentMapImpl.java:98]: Let's rename this flag to skipReserve as this is the only semantics of this flag. Also, we need to get rid of the `touch` flag in the parent call hierarchy as the naming is confusing.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/topology/PartitionsEvictManager.java [hunk 4]
```diff
@@ -111,13 +103,10 @@
      * @param grp Group context.
      */
     public void onCacheGroupStopped(CacheGroupContext grp) {
-        GroupEvictionContext grpEvictionCtx = evictionGroupsMap.remove(grp.groupId());
+        GroupEvictionContext grpEvictionCtx =
+            evictionGroupsMap.computeIfAbsent(grp.groupId(), p -> new GroupEvictionContext(grp));
```
Comment (agoncharuk) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/topology/PartitionsEvictManager.java:107]: Looks like we have a slow potential memory leak here: the value is removed from the map only on group start, so if I start and then stop unique caches, the map will grow.

Comment (ascherbakoff) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/topology/PartitionsEvictManager.java:107]: I'll create a separate ticket to resolve this.

**Pull request closed without merging**